### DISCUSSION TITLE
Rate Limiting issue

### DISCUSSION THREAD

**Discussion by and-dzh3** [emqx/emqx#4348] [Q&A]
I have a setup of emqx with this config 

  EMQX_ZONE__EXTERNAL__PUBLISH_LIMIT: 100,10s
  EMQX_LISTENER__SSL__EXTERNAL: 8883
  EMQX_STATSD__INTERVAL: "1000"
  EMQX_LISTENER__TCP__EXTERNAL__MAX_CONN_RATE: 2000
  EMQX_LISTENER__TCP__EXTERNAL__ACTIVE_N: 100

my clients for sure sending fewer packets then described in rate-limiting config 
but logs show that : [MQTT] Pause 100ms due to rate limit
and 
also tried to validate it with query the API of emqx to see how much massages are published (recv_msg) field

Msgs in 1 sec : 11
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 13
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0
Msgs in 1 sec : 0

Where the problem why clients fall into rate-limiting ?

**Comment by hjianbo** [comment]
Hi @and-dzh3 what is your emqx version?
> EMQX_ZONE__EXTERNAL__PUBLISH_LIMIT: 100,10s

means the client can send up to 100 PUBLISH messages every 10 seconds.

> Msgs in 1 sec : 11

Sorry, I'm not sure what these two numbers mean, can you explain them in detail?

The `recv_msg ` difference does not represent the rate at which this client sends messages because it is only a statistical value and is only refreshed every 15s.

**Reply by and-dzh3** [reply]
Version 4.1
Even with this configuration 
EMQX_ZONE__EXTERNAL__PUBLISH_LIMIT: 100,10s 
I receiving rate-limiting from some clients but for sure they sanding less than described rate.
As I understand from the documentation ACTIVE_N: 100 config can influence the rate-limiting mechanism but how?

I create a script that querying API each second for understanding how many packets sent in one second from some client.
thanks for the explanation about API query delay

**Reply by hjianbo** [reply]
You can get the mechanism here: https://www.emqx.io/blog/emqx_rate_limit

**Comment by mopybird** [comment]
I come across the same problem.
